src/category_theory/lifting_properties.lean

```diff
@@ -0,0 +1,208 @@
+
+/- The left lifting property of a morphism i vs. a morphism p. -/
+def has_lifting_property (i p : arrow C) : Prop := ∀ (sq : i ⟶ p), arrow.has_lift sq
+
+/- Any isomorphism has the right lifting property with respect to any map. 
+A    → X 
+↓i    ↓p≅
+B    → Y
+-/
+lemma iso_has_right_lifting_property (i : arrow C) (p : X ≅ Y) : 
+has_lifting_property i (arrow.mk p.hom) :=
+begin
+  intro sq,
+  fconstructor,
+  fconstructor,
+  fconstructor,
+  { use sq.right ≫ p.inv }, -- the lift is obtained by p⁻¹ ∘ (B → Y)
+  { rw [←category.assoc, (arrow.w sq).symm],
+    simp, },
+  { simp, }
+end
+
+/- Any identity has the right lifting property with respect to any map. -/
+lemma id_has_right_lifting_property (i : arrow C) : has_lifting_property i (arrow.mk (𝟙 X)) :=
+begin
+  haveI := is_iso.id X,
+  exact iso_has_right_lifting_property i (category_theory.as_iso (𝟙 X)),
+end
+
+/- An equivalent characterization for right lifting against a map i whose source is initial. 
+∅ → X
+↓   ↓
+B → Y has a lifting iff there is a map B → X making the right part commute.
+-/
+lemma right_lifting_property_initial_iff (i p : arrow C) 
+(h : category_theory.limits.is_initial i.left) :
+has_lifting_property i p ↔ ∀ {e : i.right ⟶ p.right}, ∃ l : i.right ⟶ p.left, l ≫ p.hom = e :=
+begin
+  split,
+  { intros hlift bottom,
+    have comm : (category_theory.limits.is_initial.to h p.left) ≫ p.hom = i.hom ≫ bottom := 
+    begin apply category_theory.limits.is_initial.hom_ext h, end,
+    haveI := hlift (arrow.hom_mk comm),
+    use (arrow.has_lift.struct (arrow.hom_mk comm)).lift,
+    rw (arrow.has_lift.struct (arrow.hom_mk comm)).fac_right,
+    refl },
+  { intros h1 sq,
+    cases h1 with e he,
+    fconstructor,
+    fconstructor,
+    fconstructor,
+    { exact e },
+    { apply category_theory.limits.is_initial.hom_ext, simp, exact h },
+    { exact he } }
+end
+
+/- A helper construction: given a square between i and f ≫ g, produce a square between i and g, 
+whose top leg uses f:
+A  → X  
+     ↓f
+↓i   Y             --> A → Y 
+     ↓g                ↓i  ↓g
+B  → Z                 B → Z 
+ -/
+def square_to_second {i : arrow C} {f : X ⟶ Y} {g : Y ⟶ Z} (sq : i ⟶ arrow.mk (f ≫ g)) :
+i ⟶ arrow.mk g :=
+{ left := sq.left ≫ f,
+  right := sq.right,
+  w' := begin
+    have : sq.left ≫ f ≫ g = i.hom ≫ sq.right := arrow.w sq,
+    rw ←category.assoc at this,
+    tidy,
+  end
+}
+
+/- The condition of having the rlp with respect to a morphism i is stable under composition-/
+lemma has_right_lifting_property_comp {i : arrow C} {f : X ⟶ Y} {g : Y ⟶ Z}
+(hf : has_lifting_property i (arrow.mk f))
+(hg : has_lifting_property i (arrow.mk g)) : 
+has_lifting_property i (arrow.mk (f ≫ g)) :=
+begin
+  intro sq0, -- a square between i and f ≫ g
+  let sq1 := square_to_second sq0, -- transform this into a square between i and g
+  
+  -- lift i vs. g
+  haveI := hg sq1,
+  have lift_structure_sq1 := arrow.has_lift.struct sq1,
+  have sq1_lift := arrow.lift sq1,
```

Comment:
`have` is for proofs; since these two things are data, you should use `let` here. Or, in this case even better: do not define these at all and use `arrow.lift sq1` directly and insteady of `lift_structure_sq1.fac_left` use `arrow.lift.fac_left sq1` directly.

Reply:
OK, I have implemented this. One question: if I remove `let d := (arrow.has_lift.struct sq2).fac_right,` and replace `rw d,` by `rw (arrow.has_lift.struct sq2).fac_right`, this does not compile. Why is that?

Reply:
I'm not exactly sure because you didn't say what the error message is, but I think that the proper way to solve this problem is to introduce a new lemma
```lean
@[simp] lemma lift.fac_right_of_to_mk {X Y : T} {f : arrow T} {g : X ⟶ Y} (sq : f ⟶ mk g)
  [has_lift sq] : lift sq ≫ g = sq.right :=
by simp_rw [←mk_hom g, lift.fac_right]
```
(and several variants of this) in `arrow.lean` and then you should be able to `rw arrow.lift.fac_right_of_to_mk`.

Reply:
If I write 
```
  { simp, tidy,
    rw ←category.assoc,
    let d := (arrow.has_lift.struct sq2).fac_right,
    let d' := (arrow.has_lift.struct sq1).fac_right,
    have : sq0.right = sq1.right := begin tidy, end,
    rw this,
    rw ←d',
    tidy,
    rw ←category.assoc,
    rw (arrow.has_lift.struct sq2).fac_right, --get an error message here!
},
```
at the end of the proof of `has_right_lifting_property_comp` (and no changes above), I get an error message in the last line: `rewrite tactic failed, did not find instance of the pattern in the target expression (arrow.has_lift.struct sq2).lift ≫ (arrow.mk f).hom`. I understand this, since the goal indeed is `((arrow.has_lift.struct sq2).lift ≫ f) ≫ g = (arrow.has_lift.struct sq1).lift ≫ g`. However, in the lean infoview, I see `d: (arrow.has_lift.struct sq2).lift ≫ f = (arrow.has_lift.struct sq1).lift := (arrow.has_lift.struct sq2).fac_right`. So why does assigning `d` some value, and then using `rw d` not produce the same as applying `rw` to the value I set `d` to?

Reply:
I have been able to resolve this at least partly.